engine/execution/ingestion/throttle.go

```diff
@@ -0,0 +1,232 @@
+package ingestion
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+// CatchUpThreshold is the number of blocks that if the execution is far behind
+// the finalization then we will only lazy load the next unexecuted finalized
+// blocks until the execution has caught up
+const CatchUpThreshold = 500
+
+// BlockThrottle is a helper struct that throttles the unexecuted blocks to be sent
+// to the block queue for execution.
+// It is useful for case when execution is falling far behind the finalization, in which case
+// we want to throttle the blocks to be sent to the block queue for fetching data to execute
+// them. Without throttle, the block queue will be flooded with blocks, and the network
+// will be flooded with requests fetching collections, and the EN might quickly run out of memory.
+type BlockThrottle struct {
+	// config
+	threshold int // catch up threshold
+
+	// state
+	mu        sync.Mutex
+	executed  uint64
+	finalized uint64
+	inited    bool
+
+	// notifier
+	processables chan<- flow.Identifier
+
+	// dependencies
+	log     zerolog.Logger
+	state   protocol.State
+	headers storage.Headers
+}
+
+func NewBlockThrottle(
+	log zerolog.Logger,
+	state protocol.State,
+	execState state.ExecutionState,
+	headers storage.Headers,
+	catchupThreshold int,
+) (*BlockThrottle, error) {
+	finalizedHead, err := state.Final().Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get finalized head: %w", err)
+	}
+
+	finalized := finalizedHead.Height
+	// TODO: implement GetHighestFinalizedExecuted for execution state when storehouse
+	// is not used
+	executed := execState.GetHighestFinalizedExecuted()
+
+	if executed > finalized {
+		return nil, fmt.Errorf("executed finalized %v is greater than finalized %v", executed, finalized)
+	}
+
+	return &BlockThrottle{
+		threshold: catchupThreshold,
+		executed:  executed,
+		finalized: finalized,
+
+		log:     log.With().Str("component", "throttle").Logger(),
+		state:   state,
+		headers: headers,
+	}, nil
+}
+
+func (c *BlockThrottle) Init(processables chan<- flow.Identifier) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.inited {
+		return fmt.Errorf("throttle already inited")
+	}
+
+	c.inited = true
+
+	var unexecuted []flow.Identifier
+	var err error
+	if caughtUp(c.executed, c.finalized, c.threshold) {
+		unexecuted, err = findAllUnexecutedBlocks(c.state, c.headers, c.executed, c.finalized)
+		if err != nil {
+			return err
+		}
+	} else {
+		unexecuted, err = findFinalized(c.state, c.headers, c.executed, c.executed+500)
```

Comment:
> c.executed+500

is 500 supposed to be `b.threshold` or just a different constant value we use for the range? If the latter, please make it a const with some comments about how it's used